src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -907,14 +923,122 @@ public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> sourc
             IPartition<TSource> partition = source as IPartition<TSource>;
             if (partition != null) return partition.Skip(count);
             IList<TSource> sourceList = source as IList<TSource>;
-            return sourceList != null ? SkipList(sourceList, count) : SkipIterator<TSource>(source, count);
+            return sourceList != null ? new SkipListIterator<TSource>(sourceList, count, int.MaxValue) : SkipIterator<TSource>(source, count);
         }
 
-        private static IEnumerable<TSource> SkipList<TSource>(IList<TSource> source, int count)
+        private sealed class SkipListIterator<TSource> : Iterator<TSource>, IPartition<TSource>
         {
-            while (count < source.Count)
+            private readonly IList<TSource> _source;
+            private readonly int _minIndex;
+            private readonly int _maxIndex;
+            private int _index;
+
+            public SkipListIterator(IList<TSource> source, int minIndex, int maxIndex)
+            {
+                Debug.Assert(source != null);
+                Debug.Assert(minIndex >= 0);
+                Debug.Assert(minIndex <= maxIndex);
+                _source = source;
+                _minIndex = minIndex;
+                _index = minIndex;
+                _maxIndex = maxIndex;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new SkipListIterator<TSource>(_source, _minIndex, _maxIndex);
+            }
+
+            public override bool MoveNext()
+            {
+                if ((state == 1 & _index <= _maxIndex) && _index < _source.Count)
+                {
+                    current = _source[_index];
```

Comment:
If we take out the `_index <= _maxIndex` part that covers `Take()` having been done, then the rest of this is an explicit form of what the `yield`ing method from #4551 does (but with one less state, because I set `_index` in the constructor), where this concern was considered already.

We could consider only-`Take` and only-`Skip` cases separately, but that seems excessive.